### ISSUE TITLE
Auto-rules: toplevel-p issues + reload

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#2895]
1. Navigate Wikipedia by clicking on links, and notice that these page loads are not seen as "toplevel", and thus auto-rules are not fired.

2. Reloads on a page without rules does not do the right thing.  Recipe:
  - Enable no-script-mode.
  - Navigate to a URL that has no rules.
  - Disable no-script-mode.
  - Reload.
  Result: `no-script-mode` gets re-enabled.  Bug is possibly in `reapply-last-active-modes` with `last-active-modes` still holding a reference to `no-script-mode`.

@aartaka ?

**Comment by Ambrevar** [comment]
The issue with `reapply-last-active-modes` is in `can-save-last-active-modes-p`: the logic assumes URL changes but does not account for reloads.

I don't understand why we can't save last-active-modes at all times anyways.  We need more comments here :)

**Comment by Ambrevar** [comment]
There may also be some faulty logic in `apply-auto-rules`:

```lisp
(defun apply-auto-rules (url buffer)
  "Apply the rules based on the URL being loaded.
Implies that the request is a top-level one."
  (let* ((rules (matching-auto-rules url buffer))
         (previous-url (previous-url buffer))
         (previous-rules (when previous-url (matching-auto-rules previous-url buffer))))
    (when (and rules previous-url (not previous-rules))
      (save-last-active-modes buffer previous-url))
    (cond
      ((not rules)
       (reapply-last-active-modes buffer))
      ((and rules (not (eq rules previous-rules)))
       (enable-matching-modes url buffer)))
    (setf (previous-url buffer) url)))
```

The first condition restores the last-active-modes, but this condition is true if we navigate from a page without rules to another page without rules.  In this case, instead of reapplying the last active modes, we should do nothing.

**Comment by aartaka** [comment]
Yes, I've noticed some rough edges too. Will try to address with a fix sometime-ish soon (:

**Comment by Ambrevar** [comment]
Thanks, let me know if I can help with anything.

**Comment by aartaka** [comment]
If you pin it down (you mostly did, I guess?) and fix it faster than myself, it'll be fine, because I'm not certain when exactly I'll get to it (:

**Comment by Ambrevar** [comment]
There are too many unknowns for me to address this (i.e. I don't know what "toplevel-p" stands for).

I think you are the right person to address this.  In the meantime, I'll send a pull request with a series of test for auto-rules.  High time we had one! :)

**Comment by aadcg** [comment]
We're all born the same but some are more big-headed than others. The plebe understands the importance of testing to ensure a sound architecture.
